{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import glob\n",
    "from functools import reduce\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/common/Atlas/' \n",
    "all_files = glob.glob(os.path.join(path,\"Tran*\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list = [pd.read_csv(open(x),delimiter=',' ,comment='#') for x in all_files]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>DesignElementAccession</th>\n",
       "      <th>Hybridization6</th>\n",
       "      <th>Hybridization4</th>\n",
       "      <th>Hybridization2</th>\n",
       "      <th>Hybridization5</th>\n",
       "      <th>Hybridization3</th>\n",
       "      <th>Hybridization1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>NAC001</td>\n",
       "      <td>261585_at</td>\n",
       "      <td>7.853361</td>\n",
       "      <td>7.876173</td>\n",
       "      <td>7.840612</td>\n",
       "      <td>7.925616</td>\n",
       "      <td>7.864773</td>\n",
       "      <td>7.431597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>NGA3</td>\n",
       "      <td>261568_at</td>\n",
       "      <td>5.065573</td>\n",
       "      <td>5.095298</td>\n",
       "      <td>5.529048</td>\n",
       "      <td>5.276014</td>\n",
       "      <td>5.231737</td>\n",
       "      <td>5.442021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>DCL1</td>\n",
       "      <td>261584_at</td>\n",
       "      <td>8.157500</td>\n",
       "      <td>7.945776</td>\n",
       "      <td>8.232055</td>\n",
       "      <td>7.667553</td>\n",
       "      <td>8.247371</td>\n",
       "      <td>7.949174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>PPA1</td>\n",
       "      <td>261579_at</td>\n",
       "      <td>10.792791</td>\n",
       "      <td>10.857476</td>\n",
       "      <td>10.832922</td>\n",
       "      <td>10.942151</td>\n",
       "      <td>10.751105</td>\n",
       "      <td>11.161513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G01060</td>\n",
       "      <td>LHY</td>\n",
       "      <td>261569_at</td>\n",
       "      <td>7.027019</td>\n",
       "      <td>7.122344</td>\n",
       "      <td>7.044126</td>\n",
       "      <td>7.299201</td>\n",
       "      <td>7.103045</td>\n",
       "      <td>7.009151</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene ID Gene Name DesignElementAccession  Hybridization6  Hybridization4  \\\n",
       "0  AT1G01010    NAC001              261585_at        7.853361        7.876173   \n",
       "1  AT1G01030      NGA3              261568_at        5.065573        5.095298   \n",
       "2  AT1G01040      DCL1              261584_at        8.157500        7.945776   \n",
       "3  AT1G01050      PPA1              261579_at       10.792791       10.857476   \n",
       "4  AT1G01060       LHY              261569_at        7.027019        7.122344   \n",
       "\n",
       "   Hybridization2  Hybridization5  Hybridization3  Hybridization1  \n",
       "0        7.840612        7.925616        7.864773        7.431597  \n",
       "1        5.529048        5.276014        5.231737        5.442021  \n",
       "2        8.232055        7.667553        8.247371        7.949174  \n",
       "3       10.832922       10.942151       10.751105       11.161513  \n",
       "4        7.044126        7.299201        7.103045        7.009151  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_list[4].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mat = reduce(lambda a,b : pd.merge(a,b,on=['Gene ID','Gene Name','DesignElementAccession'],how ='outer'),df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>DesignElementAccession</th>\n",
       "      <th>Col_BRZ_rep3</th>\n",
       "      <th>Col_mock_rep3</th>\n",
       "      <th>Col_mock_rep1</th>\n",
       "      <th>Col_mock_rep2</th>\n",
       "      <th>Col_BRZ_rep2</th>\n",
       "      <th>Col_BRZ_rep1</th>\n",
       "      <th>Sakakibara_1-2_TZ-treatment-wild_Rep2_ATH1</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM566566</th>\n",
       "      <th>GSM566556</th>\n",
       "      <th>GSM566564</th>\n",
       "      <th>GSM566569</th>\n",
       "      <th>GSM566571</th>\n",
       "      <th>GSM566573</th>\n",
       "      <th>GSM566572</th>\n",
       "      <th>GSM566561</th>\n",
       "      <th>GSM566555</th>\n",
       "      <th>GSM566557</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>NAC001</td>\n",
       "      <td>261585_at</td>\n",
       "      <td>6.360929</td>\n",
       "      <td>7.106075</td>\n",
       "      <td>7.065899</td>\n",
       "      <td>6.701490</td>\n",
       "      <td>6.390286</td>\n",
       "      <td>6.613429</td>\n",
       "      <td>4.911840</td>\n",
       "      <td>...</td>\n",
       "      <td>3.007622</td>\n",
       "      <td>3.106199</td>\n",
       "      <td>2.937103</td>\n",
       "      <td>2.982234</td>\n",
       "      <td>3.241456</td>\n",
       "      <td>3.292402</td>\n",
       "      <td>3.028502</td>\n",
       "      <td>3.031124</td>\n",
       "      <td>3.198189</td>\n",
       "      <td>3.007981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>NGA3</td>\n",
       "      <td>261568_at</td>\n",
       "      <td>4.293181</td>\n",
       "      <td>4.481487</td>\n",
       "      <td>4.399528</td>\n",
       "      <td>4.917085</td>\n",
       "      <td>4.403067</td>\n",
       "      <td>4.484703</td>\n",
       "      <td>5.331372</td>\n",
       "      <td>...</td>\n",
       "      <td>2.887678</td>\n",
       "      <td>2.918987</td>\n",
       "      <td>2.563552</td>\n",
       "      <td>2.469065</td>\n",
       "      <td>2.656620</td>\n",
       "      <td>2.677486</td>\n",
       "      <td>3.011644</td>\n",
       "      <td>2.699298</td>\n",
       "      <td>3.533549</td>\n",
       "      <td>2.450919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>DCL1</td>\n",
       "      <td>261584_at</td>\n",
       "      <td>7.418491</td>\n",
       "      <td>7.329794</td>\n",
       "      <td>7.252073</td>\n",
       "      <td>6.866045</td>\n",
       "      <td>7.178751</td>\n",
       "      <td>7.456707</td>\n",
       "      <td>7.693885</td>\n",
       "      <td>...</td>\n",
       "      <td>3.890903</td>\n",
       "      <td>2.812931</td>\n",
       "      <td>3.590816</td>\n",
       "      <td>4.431957</td>\n",
       "      <td>3.327704</td>\n",
       "      <td>3.683487</td>\n",
       "      <td>3.287549</td>\n",
       "      <td>4.121435</td>\n",
       "      <td>3.300236</td>\n",
       "      <td>3.676020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>PPA1</td>\n",
       "      <td>261579_at</td>\n",
       "      <td>10.145095</td>\n",
       "      <td>10.054270</td>\n",
       "      <td>9.911589</td>\n",
       "      <td>9.732922</td>\n",
       "      <td>9.940813</td>\n",
       "      <td>10.284046</td>\n",
       "      <td>10.227976</td>\n",
       "      <td>...</td>\n",
       "      <td>4.898910</td>\n",
       "      <td>4.385683</td>\n",
       "      <td>6.256884</td>\n",
       "      <td>6.518990</td>\n",
       "      <td>5.258337</td>\n",
       "      <td>5.115644</td>\n",
       "      <td>5.755671</td>\n",
       "      <td>6.867242</td>\n",
       "      <td>3.846946</td>\n",
       "      <td>5.840742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G01060</td>\n",
       "      <td>LHY</td>\n",
       "      <td>261569_at</td>\n",
       "      <td>9.019015</td>\n",
       "      <td>8.791247</td>\n",
       "      <td>9.067491</td>\n",
       "      <td>8.676478</td>\n",
       "      <td>9.437800</td>\n",
       "      <td>9.251234</td>\n",
       "      <td>9.975793</td>\n",
       "      <td>...</td>\n",
       "      <td>2.675125</td>\n",
       "      <td>2.851609</td>\n",
       "      <td>2.717234</td>\n",
       "      <td>2.687458</td>\n",
       "      <td>2.722341</td>\n",
       "      <td>2.752851</td>\n",
       "      <td>2.504333</td>\n",
       "      <td>2.773233</td>\n",
       "      <td>2.912970</td>\n",
       "      <td>2.717234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 234 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene ID Gene Name DesignElementAccession  Col_BRZ_rep3  Col_mock_rep3  \\\n",
       "0  AT1G01010    NAC001              261585_at      6.360929       7.106075   \n",
       "1  AT1G01030      NGA3              261568_at      4.293181       4.481487   \n",
       "2  AT1G01040      DCL1              261584_at      7.418491       7.329794   \n",
       "3  AT1G01050      PPA1              261579_at     10.145095      10.054270   \n",
       "4  AT1G01060       LHY              261569_at      9.019015       8.791247   \n",
       "\n",
       "   Col_mock_rep1  Col_mock_rep2  Col_BRZ_rep2  Col_BRZ_rep1  \\\n",
       "0       7.065899       6.701490      6.390286      6.613429   \n",
       "1       4.399528       4.917085      4.403067      4.484703   \n",
       "2       7.252073       6.866045      7.178751      7.456707   \n",
       "3       9.911589       9.732922      9.940813     10.284046   \n",
       "4       9.067491       8.676478      9.437800      9.251234   \n",
       "\n",
       "   Sakakibara_1-2_TZ-treatment-wild_Rep2_ATH1    ...      GSM566566  \\\n",
       "0                                    4.911840    ...       3.007622   \n",
       "1                                    5.331372    ...       2.887678   \n",
       "2                                    7.693885    ...       3.890903   \n",
       "3                                   10.227976    ...       4.898910   \n",
       "4                                    9.975793    ...       2.675125   \n",
       "\n",
       "   GSM566556  GSM566564  GSM566569  GSM566571  GSM566573  GSM566572  \\\n",
       "0   3.106199   2.937103   2.982234   3.241456   3.292402   3.028502   \n",
       "1   2.918987   2.563552   2.469065   2.656620   2.677486   3.011644   \n",
       "2   2.812931   3.590816   4.431957   3.327704   3.683487   3.287549   \n",
       "3   4.385683   6.256884   6.518990   5.258337   5.115644   5.755671   \n",
       "4   2.851609   2.717234   2.687458   2.722341   2.752851   2.504333   \n",
       "\n",
       "   GSM566561  GSM566555  GSM566557  \n",
       "0   3.031124   3.198189   3.007981  \n",
       "1   2.699298   3.533549   2.450919  \n",
       "2   4.121435   3.300236   3.676020  \n",
       "3   6.867242   3.846946   5.840742  \n",
       "4   2.773233   2.912970   2.717234  \n",
       "\n",
       "[5 rows x 234 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mat.to_pickle('./auxin.exp.matrix.df.pk')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/common/Atlas/Transcription profiling by array of Arabidopsis mutant for arf2 after treatment with brassinazole.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for arr10 and arr12 after treatment with t-zeatin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis after treatment with auxin and brassinazole or auxin only.csv',\n",
       " '/common/Atlas/Transcription profiling by high throughput sequencing of Arabidopsis shoots and roots in response to cytokinin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of four distinct tissues of Arabidopsis root and whole root treated with (auxin) indole-3-acetic acid against mock-treated controls.csv',\n",
       " '/common/Atlas/Transcription profiling of Arabidopsis seedlings treated with glucose and indole-3-acetic acid (auxin) to study glucose-auxin interactions.csv',\n",
       " '/common/Atlas/Transcription profiling by high throughput sequencing of Arabidopsis mutants ros1-3 and ros1-4 treated with ABA.csv',\n",
       " '/common/Atlas/Transcription profiling of Arabidopsis plants treated with indole-3-acetic acid (IAA) or IAA plus dexamethasone.csv',\n",
       " '/common/Atlas/Transcription profiling of aerial parts of Arabidopsis wild type and arr10 arr12 double mutant seedlings treated with the cytokinin trans-zeatin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis expressing miR393 or AFB1 under the control of the 35S promoter after treatment with auxin andor flg22.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis whole embryos and endosperm after treatment with abscisic acid or paclobutrazol.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis overexpressing arr22 after treatment with t-zeatin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for agb1 and or gpa1 after treatment with abscisic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis seedlings treated with brassinolide.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for sdh1 after treatment with salicylic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis Col-0, Fei-0, Bur-0, C24, Sha, Bay-0 and Bl-1 after treatment with indole-3-acetic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis T87 cells after treatment with abscisic acid and dimethylthiourea.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for nph4 or arf19 or both after treatment with indole-3-acetic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis stem tissues after treatment with naphthaleneacetic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by high throughput sequencing of Arabidopsis seedlings treated for 2 hour with cytokinin.csv']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_tpm_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " '/usr/lib/python35.zip',\n",
       " '/usr/lib/python3.5',\n",
       " '/usr/lib/python3.5/plat-x86_64-linux-gnu',\n",
       " '/usr/lib/python3.5/lib-dynload',\n",
       " '/usr/local/lib/python3.5/dist-packages',\n",
       " '/usr/lib/python3/dist-packages',\n",
       " '/usr/local/lib/python3.5/dist-packages/IPython/extensions',\n",
       " '/home/plantgenomics/.ipython',\n",
       " '/common/Atlas']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sys.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('/common/Atlas')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/common/Atlas/' \n",
    "all_files = glob.glob(os.path.join(path,\"Tran*\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/common/Atlas/Transcription profiling by array of Arabidopsis mutant for arf2 after treatment with brassinazole.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for arr10 and arr12 after treatment with t-zeatin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis after treatment with auxin and brassinazole or auxin only.csv',\n",
       " '/common/Atlas/Transcription profiling by high throughput sequencing of Arabidopsis shoots and roots in response to cytokinin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of four distinct tissues of Arabidopsis root and whole root treated with (auxin) indole-3-acetic acid against mock-treated controls.csv',\n",
       " '/common/Atlas/Transcription profiling of Arabidopsis seedlings treated with glucose and indole-3-acetic acid (auxin) to study glucose-auxin interactions.csv',\n",
       " '/common/Atlas/Transcription profiling by high throughput sequencing of Arabidopsis mutants ros1-3 and ros1-4 treated with ABA.csv',\n",
       " '/common/Atlas/Transcription profiling of Arabidopsis plants treated with indole-3-acetic acid (IAA) or IAA plus dexamethasone.csv',\n",
       " '/common/Atlas/Transcription profiling of aerial parts of Arabidopsis wild type and arr10 arr12 double mutant seedlings treated with the cytokinin trans-zeatin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis expressing miR393 or AFB1 under the control of the 35S promoter after treatment with auxin andor flg22.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis whole embryos and endosperm after treatment with abscisic acid or paclobutrazol.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis overexpressing arr22 after treatment with t-zeatin.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for agb1 and or gpa1 after treatment with abscisic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis seedlings treated with brassinolide.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for sdh1 after treatment with salicylic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis Col-0, Fei-0, Bur-0, C24, Sha, Bay-0 and Bl-1 after treatment with indole-3-acetic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis T87 cells after treatment with abscisic acid and dimethylthiourea.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis mutant for nph4 or arf19 or both after treatment with indole-3-acetic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by array of Arabidopsis stem tissues after treatment with naphthaleneacetic acid.csv',\n",
       " '/common/Atlas/Transcription profiling by high throughput sequencing of Arabidopsis seedlings treated for 2 hour with cytokinin.csv']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
